### ISSUE TITLE
[EKS] [request]: Support CoreDNS configuration

### ISSUE THREAD

**Issue by andrewpurdin** [aws/containers-roadmap#1275]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
We would like to be able to customize the CoreDNS configuration without having to hand edit the default configmap that is created for it. Although editing the configmap is a valid approach in some cases, it is an anti-pattern for infrastructure-as-code deployments since the resource is not managed in IaC. 

[AKS has solved this](https://docs.microsoft.com/en-us/azure/aks/coredns-custom) by exposing an additional configmap that one can optionally create and manage which will be imported as CoreDNS configuration. This allows users to create that additional configmap and manage it under IaC post cluster creation. 

#1159 may be a prerequisite for adding this functionality, but I am not sure.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Ultimately. we simply want to manage CoreDNS configuration with our IaC tooling. Until we can do that, we either have to avoid updating the CoreDNS configuration, or come up with specialized configurations to handle this case outside of IaC. 

**Are you currently working around this issue?**
We currently don't have a good workaround for this issue.

**Comment by bengaywins** [comment]
@andrewpurdin Why can't you just use Kustomize to do this exact thing? It would be IaC in declarative form that allows for you to run `kubectl apply -k /folder/to/your/kustomize/overlays`

**Comment by andrewpurdin** [comment]
> @andrewpurdin Why can't you just use Kustomize to do this exact thing? It would be IaC in declarative form that allows for you to run `kubectl apply -k /folder/to/your/kustomize/overlays`

I've not used Kustomize personally, but we manage the state of our EKS clusters with our IaC tool, and so any changes made via other sources will be overwritten when we re-run the pipeline related to that cluster since they aren't declared in the configuration.

**Comment by andrewpurdin** [comment]
> @andrewpurdin Why can't you just use Kustomize to do this exact thing? It would be IaC in declarative form that allows for you to run `kubectl apply -k /folder/to/your/kustomize/overlays`

I've not used Kustomize personally, perhaps we could adopt the resource after creation time with something like this. However it would still be much simpler if we could manage an object at creation time rather than having to take extra steps afterward.

**Comment by bengaywins** [comment]
The thing is, `at creation time` doesn't really exist in the sense of _before_ it creates. So it's always a step after anyway. Unless they were to avoid deploying CoreDNS entirely. Which we both know is a disaster waiting to happen.

Kustomize is rapidly becoming the suggested way to deploy all things into a K8s cluster. It's a lot better than Helm and doesn't require special knowledge.

If you don't mind saying, what IaC paradigm are you using? If Terraform, there is a Kustomize [provider](https://registry.terraform.io/providers/kbst/kustomize/latest/docs) you can use. I can't say I have used it yet, but likely will soon.

**Comment by andrewpurdin** [comment]
We're using [pulumi](https://www.pulumi.com), and I hear what you're saying about creation time, however we are able to manage the auth configmap which is created during the cluster creation process so I don't see why a configmap for controlling CoreDNS couldn't be exposed in a similar fashion. @mikestef9 can you confirm if this is a feature AWS would like to support?

**Comment by bengaywins** [comment]
https://www.pulumi.com/blog/announcing-kustomize-support/

**Comment by stevehipwell** [comment]
@bengaywins I'm not sure of the relevance of  _Kustomize_ in this as all it's doing is patching an existing object but without the fine grained control of an actual IaC tool (using Kustomize from Terraform or Pulumi might be useful as part of a migration but doesn't make a lot of sense as an actual pattern). This could be achieved by multiple means but all solutions are still just patching an EKS managed object with an unknown reconciliation loop.

@andrewpurdin  here is issue https://github.com/aws/containers-roadmap/issues/923 to create a bare cluster allowing full control of everything installed.

**Comment by matti** [comment]
related #1679

**Comment by oguzhanaygn** [comment]
I'm running EKS 1.21 with managed add-ons. When my pods/nodes are starts to downscale, some `No Route to Host` errors are coming from the application pods. I'm suspecting the CoreDNS cache I can't examine it because CoreDNS doesn't allow its configmap to be reconfigured. So I'm kinda stuck because of this situation.

**Comment by Istador** [comment]
I think the only workaround for now is to migrate from managed to self-managed `coredns`. Then you can edit the config as you please, without it getting overwritten automatically.

To switch to self-managed, you can `delete` the `coredns` add-on with the `--preserve` flag. This will keep all `coredns` resources within the cluster as they are at that moment. Afterwards you can modify the configmap. (If I remember correctly, there even is a checkbox for `--preserve` when deleting via the console.)
See https://github.com/aws/containers-roadmap/issues/1415#issuecomment-904770014

Once the feature of this issue is implemented, it's very simple to just install the add-on again (make sure to backup your custom config somewhere or it might get overwritten then).

**Comment by skupjoe** [comment]
> One possible workaround for now is to migrate from managed to self-managed `coredns`. Then you can edit the config as you please, without it getting overwritten automatically.
> 
> To switch to self-managed, you can `delete` the `coredns` add-on with the `--preserve` flag. This will keep all `coredns` resources within the cluster as they are at that moment. Afterwards you can modify the configmap. (If I remember correctly, there even is a checkbox for `--preserve` when deleting via the console.) See [#1415 (comment)](https://github.com/aws/containers-roadmap/issues/1415#issuecomment-904770014)
> 
> Once the feature of this issue is implemented, it's very simple to just install the add-on again (make sure to backup your custom config somewhere or it might get overwritten then).

Right- this is the _only_ way currently to be able to update EKS add-on configmaps or deployment/daemonset manifests, due to the presence of the `eks` manager which periodically updates managedFields, thereby overriding any custom changes that you've made.

It's unfortunate that you must deregister from AWS to do this, and I'm hoping in the future there will be a better way.

**Comment by maximmold** [comment]
Since coredns is added by default but doesn't show up in the list of addons, do you have to add it and then remove it with the reserve flag?

**Comment by sriramranganathan** [comment]
Amazon EKS team recently announced the general availability of advanced configuration feature for managed add-ons. You can now pass in advanced configuration for cluster add-ons, enabling you to customize add-on properties not handled by default settings. Configuration can be applied to add-ons either during cluster creation or at any time after the cluster is created.

Custom configuration will also allow you to customize the CoreDNS addon by allowing modification of the corefile.

To learn more about this feature, check out this blogpost - https://aws.amazon.com/blogs/containers/amazon-eks-add-ons-advanced-configuration/

Check out the Amazon EKS documentation - https://docs.aws.amazon.com/eks/latest/userguide/managing-add-ons.html

**Comment by pcebul** [comment]
Hey @sriramranganathan @mikestef9 

Sorry to bother you, but I cant find a single example on the internet... using `aws eks describe-addon-configuration` I can see under `configurationSchema` that I should be able to deploy CoreDNS addon with my custom nodeSelector, and I want to do this using CloudFormation template. But like I mentioned, I cant find a single example on web how to use ConfigurationValues field (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-addon.html#cfn-eks-addon-configurationvalues) . Also, how can I check what are the defaults for managed fields (before deploying it), eg what are the defaults for tolerations per specific addon version?
Thanks!

**Comment by sriramranganathan** [comment]
Hi @pcebul . The [blogpost](https://aws.amazon.com/blogs/containers/amazon-eks-add-ons-advanced-configuration/) should help you get started with using configurationValues when creating or updating an addon. While the blog does not show the implementation from AWS CloudFormation, the approach is similar wherein you can pass in configurationValues as a String value. Today, the configuration schema for CoreDNS addon does not support specifying tolerations but this is something we can enhance and as part of the schema, show the default values as well. The JSON schema has provision to show the default values for different parameters.

**Comment by antyung88** [comment]
`When you launch an Amazon EKS cluster with at least one node, two replicas of the CoreDNS image are deployed by default, regardless of the number of nodes deployed in your cluster.`
https://docs.aws.amazon.com/eks/latest/userguide/managing-coredns.html

coredns gets stuck provisioning through terraform, remains degraded until manually scaling up
```kubectl scale deployments.apps -n kube-system coredns --replicas=*```

**Comment by antyung88** [comment]
`When you launch an Amazon EKS cluster with at least one node, two replicas of the CoreDNS image are deployed by default, regardless of the number of nodes deployed in your cluster. `
https://docs.aws.amazon.com/eks/latest/userguide/managing-coredns.html

`This functionality has been released in [v4.47.0 of the Terraform AWS Provider](https://github.com/hashicorp/terraform-provider-aws/blob/v4.47.0/CHANGELOG.md).`
https://github.com/hashicorp/terraform-provider-aws/issues/28287

configure `cluster_addons`
```
  cluster_addons = {
    kube-proxy = {
      most_recent = true
    }
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
      most_recent       = true
    }
    coredns = {
      resolve_conflicts    = "OVERWRITE"
      most_recent          = true
      configuration_values = "{\"replicaCount\":${var.desired_size}}"
    }
  }
```

output:

```
 # module.eks.aws_eks_addon.this["coredns"] will be created
  + resource "aws_eks_addon" "this" {
      + addon_name           = "coredns"
      + addon_version        = (known after apply)
      + arn                  = (known after apply)
      + cluster_name         = "eks"
      + configuration_values = jsonencode(
            {
              + replicaCount = 3
            }
        )
```
